File: Day_11_Monkey_in_the_Middle/Python/main.py
import re
import heapq
from math import lcm
from typing import List, Tuple, Callable


# class representing monkey
class Monkey:
    def __init__(
        self,
        items: List[int],
        operation: Callable[[int], int],
        divisor: int,
        divisible_throw_to: int,
        indivisible_throw_to: int,
    ):
        # worry levels of items held by monkey
        self.items = items
        # operation to be performed on each item
        self.operation = operation
        # divisor to check divisibility for on each item
        self.divisor = divisor
        # monkeys to throw items to if divisible or not divisible
        self.throw_to = {
            True: divisible_throw_to,
            False: indivisible_throw_to,
        }
        # number of inspected items
        self.inspection_count = 0

    def inspect(self, worry_manager: Callable[[int], int]) -> List[Tuple[int, int]]:
        """
        Inspect current items and return list of monkeys to throw items to.
        """
        throw_tos = []

        for item_worry in self.items:
            # perform operation on item worry level
            item_worry = self.operation(item_worry)
            # perform worry management operation on item worry level
            item_worry = worry_manager(item_worry)
            # add to list of tuples that dictate which monkey to throw item to
            throw_tos.append((self.throw_to[item_worry % self.divisor == 0], item_worry))

        n_items = len(self.items)
        # remove inspected items
        for _ in range(n_items):
            self.items.pop(0)

        # increment inspection count
        self.inspection_count += n_items

        return throw_tos


# class for managing list of monkeys
class MonkeysManager:
    def __init__(self):
        # list of monkeys
        self.monkeys = []
        # worry management function
        self.worry_manager = lambda x: x

    def add_monkey(self, monkey: Monkey):
        """
        Add monkey to list.
        """
        self.monkeys.append(monkey)

    def set_worry_manager(self, worry_manager: Callable[[int], int]):
        """
        Set worry management function.
        """
        self.worry_manager = worry_manager

    def run_inspection_round(self):
        """
        Have each monkey inspect items sequentially.
        """
        for monkey in self.monkeys:
            throw_tos = monkey.inspect(worry_manager=self.worry_manager)
            # throw items to other monkeys
            for i, item_worry in throw_tos:
                self.monkeys[i].items.append(item_worry)

    def monkey_business(self, k: int) -> int:
        """
        Compute monkey business using product of top-k number of items inspected.
        """
        # top-k number of items inspected
        top_k_inspection_count = []
        for monkey in self.monkeys:
            # push inspection count to heap while maintaining top-k counts
            if len(top_k_inspection_count) < k:
                heapq.heappush(top_k_inspection_count, monkey.inspection_count)
            else:
                heapq.heappushpop(top_k_inspection_count, monkey.inspection_count)

        # compute product of top-k number of items inspected
        inspections_product = 1
        for i in top_k_inspection_count:
            inspections_product *= i

        return inspections_product


if __name__ == '__main__':
    # file with monkeys input data
    file_path = '../monkeys.txt'
    # open and read file
    with open(file_path) as f:
        file_contents = f.read()

    # regex pattern for parsing starting items
    regex_starting_items = re.compile(r'^\s*Starting items:\s*(.+)$')
    # regex pattern for parsing operation
    regex_operation = re.compile(r'^\s*Operation:\s*new\s*=\s*(.+)$')
    # regex pattern for parsing divisor
    regex_divisor = re.compile(r'^\s*Test:\s*divisible\s*by\s*(\d+)$')
    # regex pattern for parsing monkey to throw to if divisible
    regex_divisible_throw_to = re.compile(r'^\s*If\s*true:\s*throw\s*to\s*monkey\s*(\d+)$')
    # regex pattern for parsing monkey to throw to if indivisible
    regex_indivisible_throw_to = re.compile(r'^\s*If\s*false:\s*throw\s*to\s*monkey\s*(\d+)$')

    monkeys_manager = MonkeysManager()
    # list of divisors
    divisors = []

    for monkey_input in file_contents.split('\n\n'):
        lines = monkey_input.splitlines()

        # parse monkey infos
        result = regex_starting_items.search(lines[1])
        starting_items = [int(i.strip()) for i in result.group(1).split(',')]

        result = regex_operation.search(lines[2])
        operation = eval(f'lambda old: {result.group(1)}')

        result = regex_divisor.search(lines[3])
        divisor = int(result.group(1))
        divisors.append(divisor)

        result = regex_divisible_throw_to.search(lines[4])
        divisible_throw_to = int(result.group(1))

        result = regex_indivisible_throw_to.search(lines[5])
        indivisible_throw_to = int(result.group(1))

        # create and add monkey to monkeys manager
        monkey = Monkey(
            items=starting_items,
            operation=operation,
            divisor=divisor,
            divisible_throw_to=divisible_throw_to,
            indivisible_throw_to=indivisible_throw_to,
        )
        monkeys_manager.add_monkey(monkey)

    # compute worry management function
    # worry_manager = lambda x: x // 3
    divisors_lcm = lcm(*divisors)
    worry_manager = lambda x: x % divisors_lcm
    monkeys_manager.set_worry_manager(worry_manager)

    # run inspection rounds
    n_rounds = 10000
    for r in range(n_rounds):
        monkeys_manager.run_inspection_round()

    # compute and print monkey business using top-2 inspection counts
    print(monkeys_manager.monkey_business(2))

File: Day_11_Monkey_in_the_Middle/Go/main.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

// regex pattern for parsing starting items
var ReStartingItems = regexp.MustCompile(`^\s*Starting items:\s*(.+)$`)

// regex pattern for parsing operation
var ReOperation = regexp.MustCompile(`^\s*Operation:\s*new\s*=\s*old\s*(\*|\+)\s*(.+)\s*$`)

// regex pattern for parsing divisor
var ReDivisor = regexp.MustCompile(`^\s*Test:\s*divisible\s*by\s*(\d+)$`)

// regex pattern for parsing monkey to throw to if divisible
var ReDivisibleThrowTo = regexp.MustCompile(`^\s*If\s*true:\s*throw\s*to\s*monkey\s*(\d+)$`)

// regex pattern for parsing monkey to throw to if indivisible
var ReIndivisibleThrowTo = regexp.MustCompile(`^\s*If\s*false:\s*throw\s*to\s*monkey\s*(\d+)$`)

// function type for basic math operators
type OperatorFunc func(a int, b int) int

// operator character to operator function map
var OperatorFuncMap = map[string]OperatorFunc{
	"+": func(a int, b int) int { return a + b },
	"-": func(a int, b int) int { return a - b },
	"*": func(a int, b int) int { return a * b },
	"/": func(a int, b int) int { return a / b },
}

// function type for operations on worry value
type OperationFunc func(old int) int

// struct representing monkey
type Monkey struct {
	// worry values for items
	Items []int
	// operation to perform on worry value
	Operation OperationFunc
	// divisor to check divisibility for on each item
	Divisor int
	// number of inspected items
	InspectionCount int
}

// inspect current items for monkey and throw to other monkeys
func (monkey *Monkey) Inspect(worryManager OperationFunc, monkeyThrowToDivisible *Monkey, monkeyThrowToIndivisible *Monkey) {
	for _, worryValue := range monkey.Items {
		// perform operation on item worry level
		newWorryValue := monkey.Operation(worryValue)
		// perform worry management operation on item worry level
		newWorryValue = worryManager(newWorryValue)

		// check if divisible by divisor and through to another monkey
		if newWorryValue%monkey.Divisor == 0 {
			monkeyThrowToDivisible.Items = append(monkeyThrowToDivisible.Items, newWorryValue)
		} else {
			monkeyThrowToIndivisible.Items = append(monkeyThrowToIndivisible.Items, newWorryValue)
		}

		// increment inspection count for monkey
		monkey.InspectionCount++
	}

	// clear items for monkey
	monkey.Items = monkey.Items[:0]
}

// compute greatest common divisor of two integers
func gcd(a int, b int) int {
	if a == 0 {
		return b
	}

	if b == 0 {
		return a
	}

	return gcd(b, a%b)
}

// compute greatest common divisor of slice of integers
func GCD(a []int) int {
	v := a[0]
	for i := 1; i < len(a); i++ {
		v = gcd(v, a[i])
	}

	return v
}

// compute lowest common multiple of two integers
func lcm(a int, b int) int {
	return (a * b) / gcd(a, b)
}

// compute lowest common multiple of slice of integers
func LCM(a []int) int {
	v := a[0]
	for i := 1; i < len(a); i++ {
		v = lcm(v, a[i])
	}

	return v
}

// insert integer into sorted array and shift elements to the right
func InsertSorted(a []int, v int) {
	// don't insert if new integer is less than lowest value
	n := len(a)
	if v < a[n-1] {
		return
	}

	// perform linear scan to find where new integer should be inserted
	insert_idx := -1
	for i := range a {
		if a[i] > v {
			continue
		}

		insert_idx = i
		break
	}

	// shift slice to the right after insertion index
	copy(a[insert_idx+1:n], a[insert_idx:n-1])
	// insert new integer
	a[insert_idx] = v
}

func main() {
	// file with monkeys input data
	filePath := "../monkeys.txt"

	// read input file into bytes slice
	fileBytes, err := os.ReadFile(filePath)
	if err != nil {
		fmt.Printf("failed to read file %s, %s\n", filePath, err.Error())
		return
	}

	// get file contents string
	fileString := string(fileBytes)

	// unparsed input for each monkey
	monkeyInputs := strings.Split(fileString, "\n\n")
	// number of monkeys
	nMonkeys := len(monkeyInputs)
	// initialize slice of monkeys
	monkeys := make([]*Monkey, nMonkeys)
	// initialize slice of maps with indexes of monkeys to throw to
	throwToMap := make([]map[bool]int, nMonkeys)
	// initialize slice of divisors
	divisors := make([]int, nMonkeys)

	// iterate over monkey input
	for i, monkeyInput := range monkeyInputs {
		// create monkey struct
		monkey := &Monkey{}
		// set inspection count to zero
		monkey.InspectionCount = 0

		inputLines := strings.Split(monkeyInput, "\n")

		// parse and set starting items for monkey
		result := ReStartingItems.FindStringSubmatch(inputLines[1])
		itemsStrings := strings.Split((result[1]), ",")

		monkey.Items = make([]int, len(itemsStrings))
		for i := range itemsStrings {
			monkey.Items[i], err = strconv.Atoi(strings.TrimSpace(itemsStrings[i]))
			if err != nil {
				fmt.Printf("failed to parse %s into integer, %s\n", itemsStrings[i], err.Error())
				return
			}
		}

		// parse operation for monkey
		result = ReOperation.FindStringSubmatch(inputLines[2])
		operator := result[1]
		operand := result[2]

		// get operator function based on operator symbol
		operatorFunc, ok := OperatorFuncMap[operator]
		if !ok {
			fmt.Printf("unknown operator %s encountered\n", operator)
			return
		}

		// if operand is self
		if operand == "old" {
			// set operation function with self
			monkey.Operation = func(old int) int {
				return operatorFunc(old, old)
			}
			// otherwise expect operand to be integer
		} else {
			operatorValue, err := strconv.Atoi(operand)
			if err != nil {
				fmt.Printf("failed to parse %s into integer\n", operand)
				return
			}

			// set operation function with parsed operand
			monkey.Operation = func(old int) int {
				return operatorFunc(old, operatorValue)
			}
		}

		// parse and set divisor
		result = ReDivisor.FindStringSubmatch(inputLines[3])
		divisor, err := strconv.Atoi(result[1])
		if err != nil {
			fmt.Printf("failed to parse %s into integer\n", result[1])
			return
		}
		monkey.Divisor = divisor
		divisors[i] = divisor

		// set monkey in slice of monkeys
		monkeys[i] = monkey

		// initialize throw to map
		throwToMap[i] = make(map[bool]int)

		// parse and set monkey index to throw to if divisible
		result = ReDivisibleThrowTo.FindStringSubmatch(inputLines[4])
		divisibleThrowTo, err := strconv.Atoi(result[1])
		if err != nil {
			fmt.Printf("failed to parse %s into integer\n", result[1])
			return
		}
		throwToMap[i][true] = divisibleThrowTo

		// parse and set monkey index to throw to if indivisible
		result = ReIndivisibleThrowTo.FindStringSubmatch(inputLines[5])
		indivisibleThrowTo, err := strconv.Atoi(result[1])
		if err != nil {
			fmt.Printf("failed to parse %s into integer\n", result[1])
			return
		}
		throwToMap[i][false] = indivisibleThrowTo
	}

	// get lcm of divisors
	divisorsLCM := LCM(divisors)
	// worryManager := func(old int) int {
	// 	return old / 3
	// }
	worryManager := func(old int) int {
		return old % divisorsLCM
	}

	// perform rounds of inspections
	// nRounds := 20
	nRounds := 10000
	for i := 0; i < nRounds; i++ {
		for j := range monkeys {
			monkeys[j].Inspect(
				worryManager,
				monkeys[throwToMap[j][true]],
				monkeys[throwToMap[j][false]],
			)
		}
	}

	// get monkey business value from top-2 inspection counts
	inspectionCounts := make([]int, 2)
	for j := range monkeys {
		InsertSorted(inspectionCounts, monkeys[j].InspectionCount)
	}

	fmt.Println(inspectionCounts[0] * inspectionCounts[1])
}
